src/man/man3/wheel.3
.TH wheel 3 "29 May 2002" "HulkRobot" \" -*- nroff -*-
.ad l
.nh
.SH NAME
wheel \- 
.SH SYNOPSIS
.br
.PP
\fC#include <wheel.h>\fR
.PP
.SS Public Methods

.in +1c
.ti -1c
.RI "\fBwheel\fR (int samplingPeriod, \fBgyro\fR *aG)"
.br
.ti -1c
.RI "virtual \fB~wheel\fR ()"
.br
.ti -1c
.RI "\fBwheel\fR (\fBgyro\fR *aG, int samplingPeriod)"
.br
.ti -1c
.RI "virtual \fB~wheel\fR ()"
.br
.in -1c
.SS Static Public Methods

.in +1c
.ti -1c
.RI "int \fBsetSpeed\fR (float _targetSpeed)"
.br
.ti -1c
.RI "int \fBsetTurnSpeed\fR (float _targetTurn)"
.br
.ti -1c
.RI "int \fBgetSteerHeading\fR ()"
.br
.ti -1c
.RI "int \fBsetSteerHeading\fR (int _targetTurnDegree, float _targetTurn, float _targetSpeed=0.0)"
.br
.ti -1c
.RI "int \fBsetSpeed\fR (float _targetSpeed)"
.br
.ti -1c
.RI "int \fBsetTurnSpeed\fR (float _targetTurn)"
.br
.ti -1c
.RI "int \fBsetSteerHeading\fR (int _targetTurnDegree, float _targetTurn, float _targetSpeed=0.0)"
.br
.in -1c
.SH CONSTRUCTOR & DESTRUCTOR DOCUMENTATION
.PP 
.SS wheel::wheel (int samplingPeriod, \fBgyro\fR * aG)
.PP
Definition at line 54 of file wheel.cc.
.PP
Referenced by SpeedScaler().
.PP
.nf
00055 {
00056   aGyro = aG;
00057   dspInit();
00058   speed_command(0,0);
00059   speed_command(1,0);
00060 
00061   maxSpeedDiff            = MaxAcc*(float)samplingPeriod/1000.0;
00062   maxTurnDiff             = MaxRAcc*(float)samplingPeriod/1000.0;
00063   Error2(ErrorDebug, "maxSpeedDiff=%f, maxTurnDiff=%f",
00064                   maxSpeedDiff, maxTurnDiff);
00065   lastSpeed               = 0.0;
00066   lastTurn                = 0.0;
00067   targetSpeed   = 0;
00068   targetTurn      = 0;
00069 
00070   delayTicks      = samplingPeriod/OSTickPeriod;
00071 
00072   if(numOpen==0) {
00073     createScanTask();
00074   }
00075 
.fi
.SS wheel::~wheel ()\fC [virtual]\fR
.PP
Definition at line 107 of file wheel.cc.
.PP
Referenced by SpeedScaler().
.PP
.nf
00108 {
00109   numOpen--;
00110   if(numOpen==0)
00111     terminateScanTask();
00112 
.fi
.SS wheel::wheel (\fBgyro\fR * aG, int samplingPeriod)
.PP
Definition at line 54 of file robot/robot.cc.
.PP
.nf
00055 {
00056   aGyro = aG;
00057   dspInit();
00058   speed_command(0,0);
00059   speed_command(1,0);
00060 
00061   maxSpeedDiff            = MaxAcc*(float)samplingPeriod/1000.0;
00062   maxTurnDiff             = MaxRAcc*(float)samplingPeriod/1000.0;
00063   Error2(ErrorDebug, "maxSpeedDiff=%f, maxTurnDiff=%f",
00064                   maxSpeedDiff, maxTurnDiff);
00065   lastSpeed               = 0.0;
00066   lastTurn                = 0.0;
00067   targetSpeed   = 0;
00068   targetTurn      = 0;
00069 
00070   delayTicks      = samplingPeriod/OSTickPeriod;
00071 
00072   if(numOpen==0) {
00073     createScanTask();
00074   }
00075 
.fi
.SS wheel::~wheel ()\fC [virtual]\fR
.PP
Definition at line 107 of file wheel.cc.
.PP
.nf
00108 {
00109   numOpen--;
00110   if(numOpen==0)
00111     terminateScanTask();
00112 
.fi
.SH MEMBER FUNCTION DOCUMENTATION
.PP 
.SS int wheel::getSteerHeading (void)\fC [static]\fR
.PP
Definition at line 148 of file wheel.cc.
.PP
.nf
00149 {
00150   return aGyro->getSteerHeading();
00151 
.fi
.SS int wheel::setSpeed (float _targetSpeed)\fC [static]\fR
.PP
Definition at line 114 of file wheel.cc.
.PP
.nf
00115 {
00116   targetSpeed = _targetSpeed;
00117 
.fi
.SS int wheel::setSpeed (float _targetSpeed)\fC [static]\fR
.PP
Definition at line 114 of file wheel.cc.
.PP
Referenced by SpeedScaler(), runTest(), and turnTest().
.PP
.nf
00115 {
00116   targetSpeed = _targetSpeed;
00117 
.fi
.SS int wheel::setSteerHeading (int _targetTurnDegree, float _targetTurn, float _targetSpeed = 0.0)\fC [static]\fR
.PP
Definition at line 126 of file wheel.cc.
.PP
.nf
00128 {
00129   int initialHeading = aGyro->getSteerHeading();
00130   int finalHeading = initialHeading + _targetTurnDegree;
00131   Error2(ErrorWarning,"ini %d, final %d",initialHeading,finalHeading);
00132   int temp = initialHeading;
00133   targetSpeed = _targetSpeed;
00134   targetTurn = _targetTurn;
00135   while(temp > finalHeading)  {
00136     temp = aGyro->getSteerHeading();
00137     targetSpeed = _targetSpeed;
00138     targetTurn = _targetTurn;
00139     taskDelay(1);
00140   }
00141 
.fi
.SS int wheel::setSteerHeading (int _targetTurnDegree, float _targetTurn, float _targetSpeed = 0.0)\fC [static]\fR
.PP
Definition at line 126 of file wheel.cc.
.PP
Referenced by SpeedScaler(), hulkmaincpp(), and runTest().
.PP
.nf
00128 {
00129   int initialHeading = aGyro->getSteerHeading();
00130   int finalHeading = initialHeading + _targetTurnDegree;
00131   Error2(ErrorWarning,"ini %d, final %d",initialHeading,finalHeading);
00132   int temp = initialHeading;
00133   targetSpeed = _targetSpeed;
00134   targetTurn = _targetTurn;
00135   while(temp > finalHeading)  {
00136     temp = aGyro->getSteerHeading();
00137     targetSpeed = _targetSpeed;
00138     targetTurn = _targetTurn;
00139     taskDelay(1);
00140   }
00141 
.fi
.SS int wheel::setTurnSpeed (float _targetTurn)\fC [static]\fR
.PP
Definition at line 120 of file wheel.cc.
.PP
.nf
00121 {
00122   targetTurn = _targetTurn;
00123 
.fi
.SS int wheel::setTurnSpeed (float _targetTurn)\fC [static]\fR
.PP
Definition at line 120 of file wheel.cc.
.PP
Referenced by SpeedScaler(), runTest(), and turnTest().
.PP
.nf
00121 {
00122   targetTurn = _targetTurn;
00123 
.fi


.SH AUTHOR
.PP 
Generated automatically by Doxygen for HulkRobot from the source code.

src/device/wheel/wheel.cc
//
// wheel.cc
//
// This module controls TMC2A TurboTek motor controller board
//
// Orignally written by Yunho Jeon
// Human Robot Project
// KIST
//
// Modified by Insop Song
//
// 1997 12 1
// ver 0.5
//

#include "dfl.h"
#include "wheel.h"

#define _MODULE_NAME_ wheel 
#define REAL_HARDWARE	// comment out this line to run without hardware
#include "wheel.h"
extern "C" {
#include "speed.h"
}

SEM_ID wheel::scanTaskSem;
int wheel::scanTaskId;
int wheel::scanTaskQuit;
int wheel::delayTicks;
int wheel::numOpen;
float wheel::targetSpeed;
float wheel::targetTurn;
gyro *wheel::aGyro;


float wheel::maxSpeedDiff;
float wheel::maxTurnDiff;
float wheel::lastSpeed;
float wheel::lastTurn;
// 
// hardware related routines
//

void dspInit(void)
{
#ifdef REAL_HARDWARE
#endif
}

//
// Wheel Class Member Functions
//

wheel::wheel(int samplingPeriod, gyro *aG)
{
  aGyro = aG;
  dspInit();
  speed_command(0,0);
  speed_command(1,0);

  maxSpeedDiff 		= MaxAcc*(float)samplingPeriod/1000.0;
  maxTurnDiff		= MaxRAcc*(float)samplingPeriod/1000.0;
  Error2(ErrorDebug, "maxSpeedDiff=%f, maxTurnDiff=%f",
			maxSpeedDiff, maxTurnDiff);
  lastSpeed		= 0.0;
  lastTurn		= 0.0;
  targetSpeed   = 0;
  targetTurn 	= 0;

  delayTicks	= samplingPeriod/OSTickPeriod;

  if(numOpen==0) {
    createScanTask();
  }
  numOpen++;
  Error( ErrorDebug, "Wheel module initializing");
}

void wheel::createScanTask()
{

  scanTaskSem = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  scanTaskQuit=0;
  scanTaskId 	= taskSpawn("wheelScanTask", 10,VX_FP_TASK, 64*1024,
				(FUNCPTR) scanTask,
				0, 0, 0, 0, 0, 
				0, 0, 0, 0, 0);
}		

void wheel::terminateScanTask()
{
  scanTaskQuit=1;
  // wait until scan task quits 
  if(semTake(scanTaskSem, OSTickPeriod*1000)==ERROR) {	// wait 1 sec
    // kill the task
    Error(ErrorWarning, "Terminating scan task failed");
    Error(ErrorWarning, "maybe wheel system is not responding?");
    if(taskDelete(scanTaskId)==ERROR) {
      Error(ErrorFatal, "Deleting scan task failed");
    }
  }
  semDelete(scanTaskSem);
  Error(ErrorDebug, "Scan task terminated" );
}


wheel::~wheel()
{
  numOpen--;
  if(numOpen==0)
    terminateScanTask();
}

int wheel::setSpeed(float _targetSpeed)
{
  targetSpeed = _targetSpeed;
}

// turn speed
int wheel::setTurnSpeed(float _targetTurn)
{
  targetTurn = _targetTurn;
}

// turn degree
int wheel::setSteerHeading(int _targetTurnDegree, float _targetTurn,
			    float _targetSpeed)
{
  int initialHeading = aGyro->getSteerHeading();
  int finalHeading = initialHeading + _targetTurnDegree;
  Error2(ErrorWarning,"ini %d, final %d",initialHeading,finalHeading);
  int temp = initialHeading;
  targetSpeed = _targetSpeed;
  targetTurn = _targetTurn;
  while(temp > finalHeading)  {
    temp = aGyro->getSteerHeading();
    targetSpeed = _targetSpeed;
    targetTurn = _targetTurn;
    taskDelay(1);
  }
  Error2(ErrorWarning,"ini %d, final %d",initialHeading,temp);
  targetSpeed = 0;
  targetTurn = 0;

}

// get steer heading
int wheel::getSteerHeading()
{
  return aGyro->getSteerHeading();
}

void wheel::scanTask(int arg1, int arg2, int arg3, int arg4, int arg5,
		    int arg6, int arg7, int arg8, int arg9, int arg10)
{
    int 	distLeft, distRight;
    float	tSpeed, tTurn;
  while(!scanTaskQuit) {
    //
    // compute target direction and speed
    //

    tSpeed = targetSpeed;
    tTurn = targetTurn;

    Error2(ErrorNone, "targetSpeed=%f, targetTurn=%f",
				    tSpeed, tTurn);

    if (fabs(tSpeed-lastSpeed)>maxSpeedDiff) {
      if (tSpeed>lastSpeed)
	tSpeed=lastSpeed+maxSpeedDiff;
      else
	tSpeed=lastSpeed-maxSpeedDiff;
    }

    if (fabs(tTurn-lastTurn)>maxTurnDiff) {
      if (tTurn>lastTurn)
	tTurn=lastTurn+maxTurnDiff;
      else
	tTurn=lastTurn-maxTurnDiff;
    }

    lastSpeed=tSpeed;
    lastTurn=tTurn;

    distLeft 	= (int)((tSpeed+tTurn)*SpeedScaler+0.5);
    distRight	= (int)((tSpeed-tTurn)*SpeedScaler+0.5);

#ifdef REAL_HARDWARE
    Error2(ErrorNone, "distRight=%d, distLeft=%d", distRight, distLeft);
    speed_command(0, -distRight);
    speed_command(1, distLeft);
#endif

    taskDelay(delayTicks);
  }
  semGive(scanTaskSem);
  Error(ErrorDebug, "wheel::fire end");
}

//
// Module Interface Functions
//

int wheelInitialize(void)
{
  dspInit();
  speed_command(0, 0);
  speed_command(1, 0);
  return 0;
}

int wheelFinish(void)
{
  return 0;
}




//////////////////////// dsp board interface functions ////////////////////
extern "C" {
struct dsp_str *dsp = (struct dsp_str *) 0xfb000000;

void
setup_parm( int axis, double kp, double ki )
{
	dsp->speed[axis].mode = SETUP;
	dsp->speed[axis].kp = (int)(kp*1000.);
	dsp->speed[axis].ki = (int)(ki*1000.);
}

void
speed_command( int axis, int speed )		/* pulse per msec */
{
	dsp->speed[axis].speed = speed;
	dsp->speed[axis].mode = COMMAND;
}

int
get_speed( int axis )
{
	return dsp->speed[axis].cur_vel;
}

int
get_position( int axis )
{
	return dsp->speed[axis].cur_pos;
}

int
speed_loop()
{
  while(1) {
  speed_command(0, -10);
  speed_command(1,10);
  taskDelay(100);
  }
}

int
speed1()
{
  while(1) {
  speed_command(0,-10);
  taskDelay(100);
  }
}

int
speed2()
{
  while(1) {
  speed_command(1,-10);
  taskDelay(100);
  }
}

int
stop()
{
  speed_command(0,0);
  speed_command(1,0);
}
} // extern "C" {
